src/Entities/Donation.php

```diff
@@ -692,7 +689,41 @@ public function getEntryType( $mode = null ) {
 		return $eintrag;
 	}
 
-	private function decodeData( $data ) {
-		return unserialize( base64_decode( $data ) );
+	/**
+	 * @since 2.0
+	 * @return array
+	 */
+	public function getDecodedData() {
+		return unserialize( base64_decode( $this->data ) );
 	}
+
+	/**
+	 * @since 2.0
+	 * @param array $data
+	 */
+	public function encodeAndSetData( array $data ) {
+		$this->data = base64_encode( serialize( $data ) );
+	}
+
+	/**
+	 * Marks the donation as cancelled by updating the required fields.
+	 * No validation is done here, as this should happen in the domain.
+	 *
+	 * @since 2.0
+	 */
+	public function cancel() {
+		$this->dtDel = date( 'Y-m-d H:i:s' );
+		$this->status = 'D';
+
+		$data = $this->getDecodedData();
+
+		$data['dt_del'] = $this->dtDel;
+		$data['status'] = $this->status;
+		$data['utoken'] = '';
```

Comment:
As far as I remember, the old application never reset the token. I don't think we should/need to do that.